### PR TITLE
test: refactor runtests

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Preparation for https://github.com/pmodels/mpich/pull/5725.

Refactor runtests and prepare for modifications to add ability to run multiple tests inside a single MPI_Init/MPI_Finalize window.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Review by raffenet — COMMENTED** [review]
Looks mostly good. A couple of comments.

File: test/mpi/runtests [hunk 1]
```diff
@@ -66,7 +69,7 @@ our $g_total_seen = 0;
 our $g_num_timeout;
 our $g_num_timeout_thresh = 5;
 
-$mpiexec = "@MPIEXEC@";    # Name of mpiexec program (including path, if necessary)
```
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/runtests:69]: I can easily add and let `runtests` to check `MPIEXEC` environment variable, which should work for the test scripts.

File: test/mpi/runtests [hunk 2]
```diff
@@ -376,73 +376,43 @@ else {
...
-
-    print "\n" if $g_opt{showprogress}; # Terminate line from progress output
-    chdir $savedir;
-}
-# ---------------------------------------------------------------------------
 # Run the programs listed in the file given as the argument. 
 # This file describes the tests in the format
 #  programname number-of-processes [ key=value ... ]
 # If the second value is not given, the default value is used.
 #
```
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/runtests:393]: Will do

File: test/mpi/runtests (hunk 1 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/runtests:69]: Actually, `configure` will set the variable in `Makefile` so `make testing` will pass the correct `mpiexec` via `runtests` command line `-mpiexec=`. So the mechanism still works. The only difference is if the user want to directly invoke `runtests` and when `mpiexec` is not in the path, then user need to manually add `-mpiexec=` option. Let me add the ability to pickup `MPIEXEC` environment, so the user have one extra option to invoke `runtests`.

File: test/mpi/runtests (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/runtests]: @raffenet Done. Added the last two commits to address your comment.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

**Review by raffenet — APPROVED** [review]

**Pull request merged**